lib/puppet/thread_local.rb

```diff
@@ -0,0 +1,7 @@
+require 'concurrent'
+
+# We want to use the pure Ruby implementation even on JRuby. If we use the Java
+# implementation of ThreadLocal, we end up leaking references to JRuby instances
+# and preventing them from being garbage collected.
+class Puppet::ThreadLocal < Concurrent::RubyThreadLocalVar
```

Comment:
Are we wrapping this just out of good practice, or do we anticipate having to add things here in the future?

Reply:
I was changing things here a lot as I worked and this made it easier, but I also think it makes our choice and intent clear in one place. Concurrent doesn't really expose this class directly (though I don't expect it to change) and I'd rather not put it all over the place.